### ISSUE TITLE
Support parsing and evaluating build arguments specified before the FROM

### ISSUE THREAD

**Issue by jonohill** [renovatebot/renovate#5626]
**What would you like Renovate to be able to do?**

Support parsing and evaluating build arguments specified before the FROM.

**Describe the solution you'd like**

As Renovate parses each line before the FROM, Renovate should read any default ARG values along the way (ARG being the only statement that can legally appear before a FROM anyway).
Where these ARG values are used as part of the FROM reference, this should be interpolated and used to check for new version availability.
If a new version is available, Renovate should propose modifying the ARG value's default.

This supports use cases such as:

```
ARG IMAGE_TAG=10.1
FROM node:${IMAGE_TAG}

ARG IMAGE_TAG
LABEL version="${IMAGE_TAG}"
```
In this example, the build arg is used so that the upstream image tag can be referred to later.

```
ARG DOCKER_REGISTRY=my.private.registry.com
FROM ${DOCKER_REGISTRY}/my-image:1.2.3
```
In this example, the build arg is used so that the private registry can be changed. This can be useful for overriding the source for local builds, or if the registry URL has to change allows quickly effecting this change across many projects by configuring CI variables.

**Describe alternatives you've considered**

The alternative would be to not use build arguments at all for FROM references. However, ARG is supported in Dockerfiles before a FROM specifically for this use (and see above for example use cases).

Parsing a Dockerfile in this way could be difficult - the [proposal for ENV variables](https://github.com/renovatebot/renovate/issues/839) could be extended to ARG. The challenge with this is that the ARG may not be a version (see second use case above) and simply needs to be interpolated into the reference.

**Additional context**

The above examples are real - and current blockers for us adding Renovate to many of our Docker images. It's a trade-off, and we may go down the route of hardcoding our references if this proves difficult to resolve.

**Comment by rarkins** [comment]
This sounds like a reasonable addition to our `dockerfile` manager, however once #5617 is merged you could work around the problem that way.

**Comment by rarkins** [comment]
Not a full solution but probably a valid workaround: https://docs.renovatebot.com/modules/manager/regex/

**Comment by dlouzan** [comment]
A similar issue (though not exactly the same), pops when using e.g. `$CI_REGISTRY` predefined variable on a gitlab-ci file, the dockerfile manager will skip any updates. Using a regex manager works but it's much more complicate to use :bow:

E.g.

```json
             "deps": [
               {
                 "skipReason": "contains-variable",
                 "replaceString": "$CI_REGISTRY/my-super/duper-image:1.0-py3.9",
                 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
                 "datasource": "docker",
                 "depType": "image",
                 "depIndex": 0,
                 "updates": []
               }
             ]
```

**Comment by fgreinacher** [comment]
> A similar issue (though not exactly the same), pops up when using e.g. the [`CI_REGISTRY` predefined variable](https://docs.gitlab.com/ee/ci/variables/predefined_variables.html) on a gitlab-ci file, the dockerfile manager will skip any updates. Using a regex manager works but it's much more complicated to setup 🙇
> 
> E.g.
> 
> ```json
>              "deps": [
>                {
>                  "skipReason": "contains-variable",
>                  "replaceString": "$CI_REGISTRY/my-super/duper-image:1.0-py3.9",
>                  "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
>                  "datasource": "docker",
>                  "depType": "image",
>                  "depIndex": 0,
>                  "updates": []
>                }
>              ]
> ```

@dlouzan @rarkins I think this could to be done similarly to https://github.com/renovatebot/renovate/pull/14734, but IMO it's different from this issue. I'd suggest we split it out and I start working on it next week or so?

**Comment by viceice** [comment]
> > A similar issue (though not exactly the same), pops up when using e.g. the [`CI_REGISTRY` predefined variable](https://docs.gitlab.com/ee/ci/variables/predefined_variables.html) on a gitlab-ci file, the dockerfile manager will skip any updates. Using a regex manager works but it's much more complicated to setup 🙇
> 
> @dlouzan @rarkins I think this could to be done similarly to #14734, but IMO it's different from this issue. I'd suggest we split it out and I start working on it next week or so?

I think this should ber added here, like we already do for dependency proxy?https://github.com/renovatebot/renovate/blob/3eec5b7d88caf00d4f725e73c5b1a82242416781/lib/modules/manager/gitlabci/utils.ts#L18-L37

**Comment by fgreinacher** [comment]
Maybe you got my wrong @rarkins, I'd like to work on the `CI_REGISTRY` support, not on this one (at the moment 😄)

> I think this should ber added here, like we already do for dependency proxy?

That looks exactly right, needs some more thought because `CI_REGISTRY` will need to be replaced by something like `docker.mygitlab.com` .

**Comment by fgreinacher** [comment]
Is this one actually solved by https://github.com/renovatebot/renovate/pull/15751?

**Comment by fgreinacher** [comment]
@rarkins @viceice I created #16130 to track support for CI_REGISTRY. Maybe you want to hide the now irrelevant comments here :)

**Comment by Churro** [comment]
@rarkins, I think this issue can be closed:

- https://github.com/renovatebot/renovate/pull/15751 added support for build arguments
- https://github.com/renovatebot/renovate/issues/16130 added support for registryAliases
